### PR TITLE
Auth connection string tests

### PR DESCRIPTION
**Pull request by rozza**
Normalized the test cases
Added yml file
Added SCRAM-SHA-256 test

SPEC-818 DRIVERS-410 DRIVERS-439

### CODE REVIEW

**Comment by xdg** [comment]
Could you please break this up?  I only want to review the SCRAM-SHA-256 bits and not have the entire delta.

**Comment by rozza** [comment]
Not planning on breaking it up, not sure its worth the time - the SCRAM-SHA-256 parts are tiny and insequential compared to the main change of adding a yaml file and simplifying the tests.

**Comment by jmikola** [comment]
>  I only want to review the SCRAM-SHA-256 bits and not have the entire delta.

Looks like SCRAM-SHA-256 is just the last couple of test cases in the `source/auth/tests/connection-string.yml`: https://github.com/mongodb/specifications/pull/286/files#diff-1b062796ee524e90d1a5cebfc1e8fc02R207

**Comment by xdg** [comment]
Those seem fine.  Didn't need my review for that.

**Comment by behackett** [comment]
@xdg @rozza what's going on with this PR?

**Comment by xdg** [comment]
I've approved the SCRAM-SHA-256 bits.

File: source/auth/tests/connection-string.json [hunk 1]
```diff
@@ -1,279 +1,274 @@
...
+      "uri": "mongodb://user%40DOMAIN.COM@localhost/?authMechanism=GSSAPI",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": "user@DOMAIN.COM",
+        "password": null,
+        "db": "$external"
+      },
+      "options": null
```
Comment (behackett) [inline source/auth/tests/connection-string.json]: To match later tests, options should be:

```
{"authmechanism": "GSSAPI"}
```

File: source/auth/tests/connection-string.json [hunk 2]
```diff
@@ -1,279 +1,274 @@
...
+      "uri": "mongodb://user%40DOMAIN.COM@localhost/foo?authMechanism=GSSAPI&authSource=$external",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": "user@DOMAIN.COM",
+        "password": null,
+        "db": "$external"
+      },
+      "options": null
```
Comment (behackett) [inline source/auth/tests/connection-string.json]: Same comment here about options and authmechanism.

**Comment by behackett** [comment]
Why do the tests that expect the parser or client to raise an exception have warning set to true?

**Comment by behackett** [comment]
One last thing. Since gssapiServiceName is deprecated, can we make the test that requires it optional (maybe add an optional key set to true)? Some drivers no longer support it and shouldn't have to add back support for a deprecated option to pass the tests.

**Comment by rozza** [comment]
Closing as SPEC-1042 was born out of the contradictions in these test cases.

**Pull request closed without merging**